ipq806x: Initial support for TP-Link Archer VR2600v

shilga opened the pull request:
This pull request will provide initial support for TP-Link Archer VR2600v. It is possible to flash the sysupgrade image via tftp. Afterwards an upgrade via sysupgrade is possible.
Wifi is not working yet, so is the DSL modem.
I also began adding this router to the OpenWRT wiki: https://wiki.openwrt.org/inbox/tp-link/tp-link_archer_vr2600v_v1

mkresin commented on target/linux/ipq806x/base-files/etc/board.d/01_leds:
```diff
@@ -45,6 +45,13 @@ nbg6817)
 	ucidef_set_led_netdev "wan" "WAN" "$board:white:internet" "eth1"
 	;;
+vr2600v)
+	ucidef_set_led_usbdev "phone" "Phone" "${board}:white:phone" "1"
```
Don't use the phone LED for usb!

shilga replied:
Actually this is wrong. I did not want to control the phone LED at all here. I am currently misusing it as status LED as I could not figure out all pins.

mkresin replied:
Meh. ```power: phone {}```. Missed that. Please don't do that. Use the alias phone.

Nevertheless, the ucidef_set_led_usbdev line can be dropped. It is wrong on multiple levels. I mean usbdev as status?

mkresin replied:
Have a look at the VR2600v GPL tarball for missing key/leds. You should find something similar to this: https://github.com/lede-project/source/pull/414#discussion_r87491038

mkresin commented on target/linux/ipq806x/base-files/etc/board.d/01_leds:
```diff
@@ -45,6 +45,13 @@ nbg6817)
 	;;
+vr2600v)
+	ucidef_set_led_usbdev "phone" "Phone" "${board}:white:phone" "1"
+	ucidef_set_led_usbdev "usb" "USB" "${board}:white:usb" "1"
```
use ucidef_set_led_usbport here. This way multiple usb ports can control one LED.

Something like:

```
ucidef_set_led_usbport "usb" "USB" "${board}:white:usb" "usb1-port1" "usb2-port1" "usb3-port1" "usb4-port1"
```

should work. Find a more detailed explaination at https://github.com/lede-project/source/pull/414/commits/db9904cb8bb38c2cd2a856e0ec7d83ebe0528d5a#discussion_r88208884

mkresin commented on target/linux/ipq806x/base-files/etc/board.d/01_leds:
```diff
@@ -45,6 +45,13 @@ nbg6817)
+vr2600v)
+	ucidef_set_led_usbdev "phone" "Phone" "${board}:white:phone" "1"
+	ucidef_set_led_usbdev "usb" "USB" "${board}:white:usb" "1"
+	ucidef_set_led_netdev "lan" "LAN" "${board}:white:lan" "br-lan"
```
use

```
ucidef_set_led_switch "lan" "LAN" "${board}:white:lan" "switch0" "0x1e"
```

Have a look at #534.

mkresin commented on target/linux/ipq806x/base-files/etc/board.d/01_leds:
```diff
@@ -45,6 +45,13 @@ nbg6817)
+	ucidef_set_led_usbdev "phone" "Phone" "${board}:white:phone" "1"
+	ucidef_set_led_usbdev "usb" "USB" "${board}:white:usb" "1"
+	ucidef_set_led_netdev "lan" "LAN" "${board}:white:lan" "br-lan"
+	ucidef_set_led_netdev "dsl" "DSL" "${board}:white:dsl" "1"
```
Drop this. I'm quite sure it doesn't work.

mkresin commented on target/linux/ipq806x/base-files/etc/board.d/01_leds:
```diff
@@ -45,6 +45,13 @@ nbg6817)
+	ucidef_set_led_usbdev "usb" "USB" "${board}:white:usb" "1"
+	ucidef_set_led_netdev "lan" "LAN" "${board}:white:lan" "br-lan"
+	ucidef_set_led_netdev "dsl" "DSL" "${board}:white:dsl" "1"
+	ucidef_set_led_default "general" "general" "${board}:white:ledgnr" "1"
```
Why do you power on this led here? This LED is already handled by diag.sh due to the led-failsafe = &ledgnr in your dts.

mkresin commented on target/linux/ipq806x/base-files/lib/upgrade/platform.sh:
```diff
@@ -3,6 +3,10 @@
 PART_NAME=firmware
 REQUIRE_IMAGE_METADATA=1
 
+get_magic_long_vr2600v() {
```
why do you need your own get_magic_long function? Everything that is not explained in the commit message (which is completely empty by the way) is considered as mistake by me.

shilga replied:
The uBoot header is skipped by 0x200.

mkresin replied:
Well, not all additions to this files. Only the additions to platform_check_image() and your get_magic_long_vr2600v functions. Please update to the latest LEDE to test a sysupgrade afterwards.

shilga replied:
I actually noticed this. But if I remove my changes my branch will no longer update, as in my branch the old check is still in place. And merging around did break a lot.

mkresin replied:
Please do:

```
$ git remote add upstream https://git.lede-project.org/source.git
$ git pull --rebase upstream master
```

To pull in the latest commits from LEDE to test the sysupgrade.

shilga replied:
I tested it before I committed the change. The metadata check was already active before this function was emptied. So I had to take care of it beforehand. So just one check was removed. Had been two before.

mkresin commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -0,0 +1,416 @@
+#include "qcom-ipq8064-v1.0.dtsi"
+#include <dt-bindings/input/input.h>
```
add a blank after #include "qcom-ipq8064-v1.0.dtsi"

<strike>Please include <dt-bindings/gpio/gpio.h> here as well and put it in front of <dt-bindings/input/input.h> to keep alphabetical order.

Use the constants GPIO_ACTIVE_LOW and GPIO_ACTIVE_HIGH afterwards in stead of 1 and 0 in the gpio parameters.</strike>

Edit: GPIO_ACTIVE_LOW/GPIO_ACTIVE_HIGH are already used. by bad.

mkresin commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -0,0 +1,416 @@
+					/* 0xf90000 config? */
+					/* 0xfaf100 MAC */
+					/* 0xfaf200 PIN? 0xfaf400 ? */
+					/* 0xfc1000 ? */
```
Yeah, what's up with them?

mkresin commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -0,0 +1,416 @@
+		wifi {
+			label = "wifi";
+			gpios = <&qcom_pinmux 49 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_WLAN>;
```
Please use KEY_RFKILL here. Have a look at [package/kernel/button-hotplug/src/button-hotplug.c](https://github.com/lede-project/source/blob/master/package/kernel/button-hotplug/src/button-hotplug.c#L75) for supported keycodes,

mkresin commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -0,0 +1,416 @@
+                ledgeneral {
+			label = "ledgeneral";
+			gpios = <&qcom_pinmux 16 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_DOLLAR>;
```
BTN_0 please

heinzek replied:
This router have 5 button. Check this:
```
#define AP148_GPIO_BTN_RESET		54
#else
#define AP148_GPIO_BTN_RESET		64
#define AP148_GPIO_BTN_LEDSW		68
#ifdef CONFIG_ARCHERVR2600V1 // UN version
#define AP148_GPIO_BTN_JUMPSTART	65
#define AP148_GPIO_BTN_WIFI_2G		67
#define AP148_GPIO_BTN_WIFI_5G		54
#else // DE	version
#define AP148_GPIO_BTN_JUMPSTART	65
#define AP148_GPIO_BTN_WIFI			54
#define AP148_GPIO_BTN_DECT			67
```

shilga replied:
I added the buttons and tested them. It's working thanks.

mkresin commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -0,0 +1,416 @@
+	gpio-leds {
+		compatible = "gpio-leds";
+
+		/* TODO: Find staus and WAN LEDs */
```
????

mkresin commented:
What means that wifi doesn't work? Would be please be a bit more detailed on that. Information what is working and what is not working need to go into the commit message.

Please have a look at #414. Especially the tp-link safeloader part. it might allow you to create an image which can be flashed from the tp-link webinterface.

@dissent1 @heinzek any opinions?

shilga commented:
Hi Mathias,

tp-link safeloader will not work as TP-Link has changed the header. It
is now binary and has a signature in it but it still somehow looks like
the TP-Link header:

|hexdump
Archer_VR2600vv1_1.3.0_0.8.0_up_boot\(160622\)_2016-06-22_10.18.36.bin
-n 512 -C||
||00000000  03 00 00 03 0b a0 a9 ba  2a 33 90 1f 22 85 0b 54 
|........*3.."..T|||
||00000010  82 65 62 7e 00 11 22 33  44 55 66 77 88 99 aa bb 
|.eb~.."3DUfw....|||
||00000020  cc dd ee ff 03 73 95 9b  dc 33 29 ee ec f8 7e b0 
|.....s...3)...~.|||
||00000030  42 e3 d2 34 01 5f dc 43  4f 00 00 00 00 00 00 00 
|B..4._.CO.......|||
||00000040  8b 9c 2f 3a 55 75 32 cb  92 45 6e 4e 55 96 7b 93 
|../:Uu2..EnNU.{.|||
||00000050  00 00 00 00 55 aa 55 aa  f1 e2 d3 c4 e5 a6 6a 5e 
|....U.U.......j^|||
||00000060  4c 3d 2e 1f aa 55 aa 55  00 80 00 c0 00 80 00 c0 
|L=...U.U........|||
||00000070  00 02 f8 00 00 02 00 00  c8 c1 1c 00 00 00 20 00 
|.............. .|||
||00000080  3c f8 97 00 00 00 00 00  00 00 32 00 03 01 aa 55 
|<.........2....U|||
||00000090  00 08 00 a5 00 00 00 00  57 ff 4e d2 ff ff ff ff 
|........W.N.....|||
||000000a0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff 
|................|||
||*||
||000000d0  c8 d3 2f b1 d9 10 e2 db  c3 de 30 95 f1 cd 3b 9d 
|../.......0...;.|||
||000000e0  a6 07 d3 ed 20 99 78 eb  5e 8f e8 97 c9 53 a5 25  |....
.x.^....S.%|||
||000000f0  53 47 ea 32 90 69 83 20  4c 36 df c4 0e e5 5e f3  |SG.2.i.
L6....^.|||
||00000100  81 fc 02 b1 04 d9 c4 d1  8a 5a 39 67 02 e9 2f 7f 
|.........Z9g../.|||
||00000110  4d ef 4e d6 10 61 b1 1a  09 4e 6b 72 8e e9 4b 8e 
|M.N..a...Nkr..K.|||
||00000120  05 11 1a ee c2 c2 07 80  4e 00 f8 77 f8 9e 27 ce 
|........N..w..'.|||
||00000130  1c 15 57 98 d8 fc b8 b7  b7 b2 45 c2 b0 7d 9d c4 
|..W.......E..}..|||
||00000140  c9 cd 02 89 b5 d3 65 ea  5e e7 ac 2a 34 52 6a 5b 
|......e.^..*4Rj[|||
||00000150  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 
|................|||
||*||
||000001d0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff 
|................|||
||*||
||00000200|

This header is also on the flash at 0x320000 (kernel partition) with the
signature. But the header is unchecked by the bootloader. It just skips
the header therefore the kernel begins at 0x320200. Because of this the
unorthodox images are necessary.

I will provide more information about all of this in the forum. Would 
that be okay?

Regards

Sebastian

On 20.11.2016 17:02, Mathias Kresin wrote:
>
> What means that wifi doesn't work? Would be please be a bit more
> detailed on that. Information what is working and what is not working
> need to go into the commit message.
>
> Please have a look at #414
> <https://github.com/lede-project/source/pull/414>. Especially the
> tp-link safeloader part. it might allow you to create an image which
> can be flashed from the tp-link webinterface.
>
> @dissent1 <https://github.com/dissent1> @heinzek
> <https://github.com/heinzek> any opinions?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lede-project/source/pull/541#issuecomment-261786680>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAzKyXVUYkM1QzU77obJ1OzwPkXp0wVfks5rAG8XgaJpZM4K3kE4>.
>

shilga commented:
I should not have replied by mail. The hexdump is better to see here:
```
hexdump Archer_VR2600vv1_1.3.0_0.8.0_up_boot\(160622\)_2016-06-22_10.18.36.bin -n 512 -C
00000000  03 00 00 03 0b a0 a9 ba  2a 33 90 1f 22 85 0b 54  |........*3.."..T|
00000010  82 65 62 7e 00 11 22 33  44 55 66 77 88 99 aa bb  |.eb~.."3DUfw....|
00000020  cc dd ee ff 03 73 95 9b  dc 33 29 ee ec f8 7e b0  |.....s...3)...~.|
00000030  42 e3 d2 34 01 5f dc 43  4f 00 00 00 00 00 00 00  |B..4._.CO.......|
00000040  8b 9c 2f 3a 55 75 32 cb  92 45 6e 4e 55 96 7b 93  |../:Uu2..EnNU.{.|
00000050  00 00 00 00 55 aa 55 aa  f1 e2 d3 c4 e5 a6 6a 5e  |....U.U.......j^|
00000060  4c 3d 2e 1f aa 55 aa 55  00 80 00 c0 00 80 00 c0  |L=...U.U........|
00000070  00 02 f8 00 00 02 00 00  c8 c1 1c 00 00 00 20 00  |.............. .|
00000080  3c f8 97 00 00 00 00 00  00 00 32 00 03 01 aa 55  |<.........2....U|
00000090  00 08 00 a5 00 00 00 00  57 ff 4e d2 ff ff ff ff  |........W.N.....|
000000a0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
000000d0  c8 d3 2f b1 d9 10 e2 db  c3 de 30 95 f1 cd 3b 9d  |../.......0...;.|
000000e0  a6 07 d3 ed 20 99 78 eb  5e 8f e8 97 c9 53 a5 25  |.... .x.^....S.%|
000000f0  53 47 ea 32 90 69 83 20  4c 36 df c4 0e e5 5e f3  |SG.2.i. L6....^.|
00000100  81 fc 02 b1 04 d9 c4 d1  8a 5a 39 67 02 e9 2f 7f  |.........Z9g../.|
00000110  4d ef 4e d6 10 61 b1 1a  09 4e 6b 72 8e e9 4b 8e  |M.N..a...Nkr..K.|
00000120  05 11 1a ee c2 c2 07 80  4e 00 f8 77 f8 9e 27 ce  |........N..w..'.|
00000130  1c 15 57 98 d8 fc b8 b7  b7 b2 45 c2 b0 7d 9d c4  |..W.......E..}..|
00000140  c9 cd 02 89 b5 d3 65 ea  5e e7 ac 2a 34 52 6a 5b  |......e.^..*4Rj[|
00000150  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
000001d0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00000200
```

mkresin commented:
Got it. The famous TP-Link v3 header with RSA signature.

Please add this information to the commit message. You know, the commit title should show what has been done. The commit message should explain why something has been done (this way). An explanation for the unusual image format had saved both of us time...

shilga commented:
Actually this is not just the v3 header. It's more a v3.3 as the magic number suggests. Also the v3 header has some human readable information in it.
I will update my commit in regard to your comments. Sorry for being inconvenient. I tried to get anything right but do not know all standards you guys have.

mkresin commented:
> Sorry for being inconvenient. I tried to get anything right but do not know all standards you guys have.

A lot of stuff you have done is fine.

It's just that a reasonable amount of board support patches send via github are either untested, full of workaround or still work in progress. And no one mentions the support state of the board or remaining issues.

It's annoying to review these patches, since it's not obvious what is bug, workaround or just not implemented.

I mean, I'm not really interested in committing work in progress stuff. As soon as the patch is committed ones come up and opens a bugreport about missing stuff...

shilga commented:
Thanks for your thoughts and review.
I updated my commit and took in the hints you provided. I did not update the button names/ids as I copied it from the C2600 and it is working there. Those routers are quite similar from their appearance but TP-Link sadly changed a lot. I thought it would be easier to port the C2600 to the VR2600v. But it is more work than just adding the xDSL and DSP support. It could be that I need some help on the Wifi. Because of this I brought it to the point there sysupgrade is possible for convenience.

heinzek commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -0,0 +1,410 @@
+			gpios = <&qcom_pinmux 66 GPIO_ACTIVE_HIGH>;
+		};
+	};
+
```
Check and add missing leds. This is from YP-Link GPL tarball:
```
#define TP_GPIO_LED_POWER				26
#define TP_GPIO_LED_DSL					7
#define TP_GPIO_LED_USB					8
#define TP_GPIO_LED_LAN					9
#define TP_GPIO_LED_2G					16
#define TP_GPIO_LED_5G					17
#define TP_GPIO_LED_ESATA				6
#define TP_GPIO_LED_WPSPNONE			53
#define TP_GPIO_LED_INTERNET			56
#define TP_GPIO_LED_ONOFF				66
#define TP_GPIO_LED_PHONE				53
```

shilga replied:
Just for info: There did you find it in the tarball? I really searched extensive in there for this pins.

heinzek replied:
i find this using grep command:
`grep -i 'vr2600' -r *`
and listen some files using gedit

shilga replied:
I added also those LEDs and most of them are working. Actually there is no eSATA LED or I could not find it, so this has no function and I did not add it. The Power LED is not working so I also missed this one out.

mkresin replied:
Not working means not GPIO controllable, right? Obviously TP-Link likes to connect the power led to VCC. I've the same problem on my TD-W8980.

mkresin replied:
@heinzek ```cat /sys/kernel/debug/gpio``` should show you all gpios in use. what do you mean with does not work. But better we discuss this per mail and do not hijack this thread. would you mind to send me a mail with the cat output and the answer to my question.

shilga replied:
That means they do not work. I just added the LEDs that I confirmed working. Actually every LED that I can see can be switched on. The PowerLED is always on and cannot be switched off.

@heinzek I hope that this monster can eventually replace my TD-W8980.

mkresin commented on target/linux/ipq806x/base-files/etc/board.d/01_leds:
```diff
@@ -45,6 +45,15 @@ nbg6817)
 	ucidef_set_led_netdev "wan" "WAN" "$board:white:internet" "eth1"
 	;;
+vr2600v)
+	ucidef_set_led_default "dsl" "DSL" "${board}:white:dsl" "0"
```
Drop all instances of ucidef_set_led_default where you switch off LEDs and tell me which LEDs are on by default albeit they should be off.

shilga replied:
All LEDs are off by default. It's just so they are added to UCI.

mkresin replied:
Personally I don't like to add LEDs just to have them available in uci. But something similar was committed this week. Therefore I do **not** resist on removing this ucidef_set_led_default.

mkresin commented on target/linux/ipq806x/base-files/etc/board.d/01_leds:
```diff
@@ -45,6 +45,15 @@ nbg6817)
 	;;
+vr2600v)
+	ucidef_set_led_default "dsl" "DSL" "${board}:white:dsl" "0"
+	ucidef_set_led_usbport "usb" "USB" "${board}:white:usb" "usb4-port1" "usb2-port1"
```
As far as I understand this covers only usb 2.0 devices attached to the usb port. Have a look at one of my earlier comments. I suggested some more usb ports to add and linked to comment where I got the informations from.

shilga replied:
I tested it with a USB3 device and it was working. There are only 2 USB ports on this device. Maybe this will not work for USB2 devices?

shilga replied:
It was not working with USB2 devices. Changed it. Now it works.

mkresin commented on target/linux/ipq806x/base-files/etc/board.d/01_leds:
```diff
@@ -45,6 +45,15 @@ nbg6817)
+	ucidef_set_led_default "wlan2g" "WLAN2g" "$board:white:wlan2g" "0"
+	ucidef_set_led_default "wlan5g" "WLAN5g" "$board:white:wlan5g" "0"
+	ucidef_set_led_switch "wan" "wan" "${board}:white:wan" "switch0" "0x20"
+	ucidef_set_led_default "general" "general" "${board}:white:ledgnr" "1"
```
What about dropping this here and use the general led like the power led on other boards?

```
+	aliases {
+		serial0 = &uart4;
+		mdio-gpio0 = &mdio0;
+
+		led-boot = &ledgnr;
+		led-failsafe = &ledgnr;
+		led-running = &ledgnr;
+		led-upgrade = &ledgnr;
+	};
```

shilga replied:
If this is working. Why not? Thought that would lead to problems.

heinzek commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -0,0 +1,431 @@
+		};
+
+		wlan5g {
+			label = "vr2600v:white:wlan2g";
```
change label to "vr2600v:white:wlan**5**g"

shilga replied:
woopsie. Yes of course.

mkresin commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -0,0 +1,431 @@
+			linux,code = <BTN_0>;
+		};
+
+                ledgeneral {
```
I guess LEDSW from the GPL tarball means ledswitch

shilga replied:
Yes. The original function is to switch off all LEDs.

mkresin replied:
Than change it to ledswitch. Check http://lxr.free-electrons.com/source/include/dt-bindings/input/linux-event-codes.h for an linux,code that matches "switch leds off". If you are not able to find something like that, keep BTN_1.

shilga replied:
I actually don't care if I can switch off the LEDs or not. They are not that bright. The button has actually no label at all. I tested the button with the original firmware only because of this I know what the TP-Link engineers thought with this. So whoever does want to do something with it can do that. Hence the general. I took the wording from the C2600. The top of those routers look the same. This is a button with a LED in it. This LED is now used as status during boot and update.

shilga replied:
Should I also change the name of the corresponding LED?

shilga replied:
I decided to keep BTN_1. There would be some events that could be used but I could not decide on which one. And BTN_1 has one advantage: It is recognized by hotplug.

mkresin replied:
> It is recognized by hotplug

Bad argument: https://git.lede-project.org/9dec029e7f14fb84ac8d751a516ab522b3255553

Please write which events could be used and I will tell you which one is right one.

shilga replied:
I suggest for the phone button KEY_PHONE which you already added in your commit from October.
For the ledswitch I suggest one of the following: KEY_BRIGHTNESS_ZERO, KEY_DISPLAY_OFF, KEY_DISPLAYTOGGLE, LED_MUTE, KEY_LIGHTS_TOGGLE 

I think the most fitting would be: KEY_LIGHTS_TOGGLE

mkresin replied:
KEY_LIGHTS_TOGGLE is perfect. Use this one.

mkresin commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -0,0 +1,431 @@
+			gpios = <&qcom_pinmux 56 GPIO_ACTIVE_HIGH>;
+		};
+
+		ledgnr: ledgnr {
```
What do you think about using ledgeneral as name. this ledgnr abbreviation is really hard to get.

mkresin replied:
Could you rename it to just general? We might have a fix for your power led as well. It is tested at the moment.

shilga commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -0,0 +1,431 @@
+					ART@fc0000 {
+						label = "ART";
+						reg = <0xfc0000 0x4000>;
+						read-only;
```
Good one. Thanks. Hoped that would fix wifi but it did not.

mkresin replied:
Please provide a full bootlog. Maybe we can spot something relevant.

shilga replied:
Bootlog can be found here: http://pastebin.com/hpwGcgz7

I want to create a topic in the forum to share all my findings and hoping to find other owners of this device.

heinzek commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -0,0 +1,431 @@
+					       "gpio46", "gpio47";
+					bias-bus-hold;
+				};
+			};
```
Remove nand pins, this router not have nand chip.

heinzek commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -0,0 +1,431 @@
+				function = "gsbi4";
+				bias-disable;
+			};
+
```
remove i2c4_pins, this pinmux is not used on this dts. Check my patch: https://github.com/lede-project/source/pull/549

heinzek commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -0,0 +1,431 @@
+			label = "vr2600v:white:wlan5g";
+			gpios = <&qcom_pinmux 17 GPIO_ACTIVE_HIGH>;
+		};
+
```
Add power led to gpio 26, and add pinmux for leds and button: https://github.com/lede-project/source/pull/549

shilga replied:
I had already seen your pull request on was already working. I am testing. What I can say is: The power LED is working now. :)

heinzek commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -0,0 +1,431 @@
+			gpio-export,name = "usb2_power";
+			gpio-export,output = <1>;
+			gpios = <&qcom_pinmux 23 GPIO_ACTIVE_HIGH>;
+		};
```
Did You check if You can control usb power using command:
power off usb 1:
echo 0 > /sys/class/gpio/usb1_power/value
power on usb 1:
echo 1 > /sys/class/gpio/usb1_power/value
power off usb 2:
echo 0 > /sys/class/gpio/usb2_power/value
power on usb 2:
echo 1 > /sys/class/gpio/usb2_power/value
add gpiomux to usbvcc:
https://github.com/lede-project/source/pull/549

shilga commented:
I did the same cleanup as in #549 to my dts file. This lead to a controllable power led. Also I checked if the USB power is actually switchable. It is not. Therefore I removed those pins.

heinzek commented:
Maybe You create topic about this router to lede forum https://forum.lede-project.org/ so we can help You to run wireless.

mkresin commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -0,0 +1,409 @@
+			linux,code = <KEY_WPS_BUTTON>;
+		};
+
+                dect {
```
spaces vs. tabs. Same applies to the ledswitch node. Would you please check this file for more space vs. tabs issues!

shilga replied:
just to be sure here: Which one to you want? tabs or spaces?

mkresin replied:
most of the file use tabs. so please use tabs as well.

shilga replied:
Thought so. Why there have somehow ended up spaces I don't know. The editor usually does that automatically. This file was based on another one, so maybe this was just copy pasted. Anyhow I fixed it.

mkresin commented on target/linux/ipq806x/base-files/lib/upgrade/platform.sh:
```diff
@@ -57,6 +57,11 @@ platform_do_upgrade() {
 		MTD_CONFIG_ARGS="-s 0x200000"
 		default_do_upgrade "$ARGV"
 		;;
+	vr2600v)
```
didn't fixed we this alphabetical distortion already? Anyway, please keep alphabetical order.

shilga replied:
No, that was one rule that I did already know. But I must have missed this one. Fixed.

heinzek commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -0,0 +1,409 @@
+			pinctrl-0 = <&rgmii2_pins>;
+			pinctrl-names = "default";
+
+			mtd-mac-address = <&defaultmac 0x8>;
```
change to <&defaultmac 0>;
Because Your mac start on 0x0 in defaultmac part

shilga commented:
I updated my commit in order to:
- fix MAC adresses
- by fixing the MAC address of Wifi it will now also come up. So Wifi is working now
- Changed I2C4 like in #549

mkresin commented:
Seams to me we are done here, right?

@shilga would you please adjust your commit message slightly . It looks a bit messy. You might want to use something like:

```
ipq806x: add support for TP-Link Archer VR2600v

This router is similar to the C2600. Ethernet on WAN + LAN, switch,
sysupgrade, LEDs, buttons and WiFi on 2G + 5G do work. The xDSL modem
is not supported yet.

It is not possible to flash LEDE via the TP-Link webinterface. The
image need to be signed. The first 0x200 bytes of the image is the
TP-Link header including the signature. The signature is not validated
by the bootloader. The LEDE image is zeroed in this area.

To install LEDE ... <insert hint how to install LEDE here>...
```

Do you know whether wireless and one of the ethernet interfaces share the mac address as well when running the TP-Link firmware?

shilga commented:
I think so, too. Now it is very identical to the C2600 from a functional view. 
I adjusted the commit message like you proposed. 

I had a look at factory firmware and can confirm that LAN and wifi0 share the same MAC. It's the one labeled on the device.

mkresin commented:
merged, thanks!
